Make `getVersion` resolve a per-project view of the `:unifiedClasspath` (#1667)
palantir/gradle-consistent-versions: docs/adr/0001-getversion-per-project-resolution.md

        at com.palantir.gradle.versions.GetVersionPlugin.getOptionalVersion(GetVersionPlugin.java:98)
        at com.palantir.gradle.versions.GetVersionPlugin.getVersion(GetVersionPlugin.java:83)
```

This affects any subproject `getVersion` call but commonly surfaces via `sls-packaging`'s `minimumVersion` usage:
```groovy
distribution {
    productDependency {
        productGroup = "com.palantir.group"
        productName = "my-service"
        minimumVersion = getVersion('com.palantir.group', 'my-service-module-api')
        maximumVersion = "1.x.x"
    }
}
```

### Approaches considered

- **Read the `versions.lock` file directly.** The `getVersion('group:name')` lookup could parse the resolved version straight out of `versions.lock`. But `GetVersionPlugin` is as part of GCV itself, it has direct access to the very same resolved dependency model GCV uses to *write* `versions.lock`, so re-parsing its own serialised output would be backwards. It also wouldn't work: the lockfile only records *external* modules, so it couldn't serve a `getVersion` lookup that resolves to another project in the same build.

> - **Read the `gcvLocks` constraints instead of resolving.** The above error only applies to *resolution* — reading a configuration's declared `DependencyConstraint`s is fine, so the lookup could read the strict constraints off the root `gcvLocks` platform instead. This works for the common case but still reads the root project's model from a subproject, it needs special-casing for the `getVersion(project(...))` case.

Reviewer: > it needs special-casing for the `getVersion(project(...))` case.

and 

> is the only one that keeps all `getVersion` overloads working without special-casing.

It's not clear what special casing you're referring to here.

Author: fixed in [03f9599](https://github.com/palantir/gradle-consistent-versions/pull/1667/commits/03f9599211f0ba6c1f7020fcd0d16e86081f060d)

- **Per-project view of the unified graph (chosen).** Each project resolves its *own* configuration instead of reaching into the root's `unifiedClasspath`. This is the isolated-projects-shaped route and is the only one that keeps all `getVersion` overloads working without special-casing.

## Decision

Each project resolves a per-project view of the unified dependency graph instead of reaching into the root's `unifiedClasspath`.

> This follows Gradle's intended separation of configuration roles. On the root project, a **dependency-scope configuration** holds the declared dependencies, and both the resolvable view that computes the locks and the consumable view that exposes them extend it. Each project then gets its own resolvable configuration that *consumes* that root view through a project dependency requesting its capability.

Reviewer: `view` -> `configuration`, they are configurations!

- **`VersionsLockPlugin`** (root): collects every project's production + test dependencies into a new dependency-scope configuration, `unifiedClasspathDependencies`. Two views extend it: `unifiedClasspath` (resolvable) still computes the lock state — behaviour is unchanged — and a new `unifiedClasspathElements` (consumable) exposes the same graph to other projects under capability `gcv:unified-classpath:0`.

> - **`GetVersionPlugin`** (applied to every project): registers a resolvable `gcvGetVersions` configuration that depends on `project(':')` requesting capability `gcv:unified-classpath:0`, and wires `getVersion(...)` to resolve that configuration.

Reviewer: > to resolve that configuration

configuration is overloaded here. 

I would say something along the lines of:

> * registers a resolvable `gcvGetVersions` configuration that effectively points to `unifiedClasspathElements` from above by depending on the root project and requiring the same capability: `gcv:unified-classpath:0`. 
> * `getVersion("some:dependency")` is then configured to resolve against `gcvGetVersions` in the _same_ instead of the `unifiedClasspath` configuration that resides in the root project.

Author: [fe80fae](https://github.com/palantir/gradle-consistent-versions/pull/1667/commits/fe80fae4f459a709173115ef3f0c3a5f00c0448a) fixed here

        elements["<b>unifiedClasspathElements</b><br/>capability gcv:unified-classpath:0"]
        unified["<b>unifiedClasspath</b><br/>→ writeLocks / verifyLocks"]
        elements -->|extendsFrom| dependencies
        unified -->|extendsFrom| dependencies
    end
    subgraph child["every subproject (including root)"]
        gv["getVersion('group:name')"]
        getv["<b>gcvGetVersions</b>"]
        gv -->|resolves| getv
    end
    getv -->|"depends on, with capability gcv:unified-classpath:0"| elements

    classDef declarable fill:#3b82f6,stroke:#1d4ed8,color:#fff
    classDef resolvable fill:#a855f7,stroke:#7e22ce,color:#fff
    classDef consumable fill:#22c55e,stroke:#15803d,color:#fff
    classDef func fill:#e5e7eb,stroke:#9ca3af,color:#111

    class dependencies declarable
    class unified,getv resolvable
    class elements consumable
    class gv func

    style root fill:none,stroke:#bbb,color:#111
    style child fill:none,stroke:#bbb,color:#111
```

Box colours by configuration role: 🟦 **blue** = dependency-scope, 🟪 **purple** = resolvable, 🟩 **green** = consumable, ⬜ **grey** = plain function.

## Consequences

> `getVersion` now resolves the calling project's own `gcvGetVersions` configuration. Via the capability-selected consumable view, that configuration sees exactly the same set of dependencies collected into `unifiedClasspathDependencies`, so the version returned is identical to before. The only difference is that the resolution happens in the calling project rather than reaching into another project's configuration, so the exclusive-lock error is gone.

Reviewer: same thing here, please drop this view nomenclature

On Gradle 9 this fails when it is called from a subproject:

```
Caused by: org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$IllegalResolutionException:
Resolution of the configuration ':unifiedClasspath' was attempted without an exclusive lock. This is unsafe and not allowed.
        at com.palantir.gradle.versions.GetVersionPlugin.getOptionalVersion(GetVersionPlugin.java:98)
        at com.palantir.gradle.versions.GetVersionPlugin.getVersion(GetVersionPlugin.java:83)
```

This affects any subproject `getVersion` call but commonly surfaces via `sls-packaging`'s `minimumVersion` usage:
```groovy
distribution {
    productDependency {
        productGroup = "com.palantir.group"
        productName = "my-service"
        minimumVersion = getVersion('com.palantir.group', 'my-service-module-api')
        maximumVersion = "1.x.x"
    }
}
```

### Approaches considered

> - **Read the `versions.lock` file directly.** The `getVersion('group:name')` lookup could parse the resolved version straight out of `versions.lock`. But `GetVersionPlugin` is as part of GCV itself, it has direct access to the very same resolved dependency model GCV uses to *write* `versions.lock`, so re-parsing its own serialised output would be backwards. It also wouldn't work: the lockfile only records *external* modules, so it couldn't serve a `getVersion` lookup that resolves to another project in the same build.

Reviewer 2: just flagging - gaia seems to do something similar to this one: https://github.palantir.build/acme/gaia/blob/8bc77e95f132498adce1c55b284ec4c8685dc411/build.gradle#L97. I am not sure if this would still work with the changes in this PR.

Author: I think this should still work with this PR as they are overriding the `getVersion` for there own

Author: I don't think this will make GCV CoD compatible so think they will still need that workaround

> # 1. `getVersion` resolves a per-project view of the unified classpath

Reviewer: please turn on line wrapping in your IDE, and have things be single line.

It makes editing and future diffs incredibly annoying to read without this.

Author: fixed in 532d8923a46ec0fea09082ea723aa8a0d78353b9

## Context

`getVersion('group:name')` (from `GetVersionPlugin`) lets build scripts look up the
version GCV settled on for a dependency. The no-configuration overloads originally

> worked by resolving the **root** project's `unifiedClasspath` configuration to read the

Reviewer: you haven't introduced the overloads, let's not talk about them, considering you've been very specific in the form that you're referring to.

```suggestion
version GCV settled on for a dependency. Originally, this
worked by resolving the **root** project's `unifiedClasspath` configuration to read the
```

version back out, regardless of which project `getVersion` was called from.

> On Gradle 9 this fails when `getVersion` is called from a subproject:

Reviewer: it's weird to put `getVersion` here because now you have overloads to think about, I would just say

```suggestion
On Gradle 9 this fails when it is called from a subproject:
```

```
Caused by: org.gradle.api.internal.artifacts.configurations.DefaultConfiguration$IllegalResolutionException:
Resolution of the configuration ':unifiedClasspath' was attempted without an exclusive lock. This is unsafe and not allowed.
        at com.palantir.gradle.versions.GetVersionPlugin.getOptionalVersion(GetVersionPlugin.java:98)
        at com.palantir.gradle.versions.GetVersionPlugin.getVersion(GetVersionPlugin.java:83)
```

This was a warning on Gradle 8 and is now an error: a subproject may not resolve the

> root project's configuration. It commonly surfaces via `sls-packaging`'s

Reviewer: > a subproject may not resolve the root project's configuration

is this supposed to be a reference to an error? or something else

Author: This was supposed to reference the above error but honestly doesn't add anything useful so removed in 1b55dd3f37d5b6b5472343fed9d9250687a108e8

`minimumVersion` usage inside `subprojects { ... }`, but affects any subproject

> `getVersion` call.

Reviewer: I would put an example of the sls-packaging usage, it is not obvious what you're referring to.

Author: example added in 1b55dd3f37d5b6b5472343fed9d9250687a108e8

### Approaches considered

- **Read the `versions.lock` file directly.** All the information is there, but reading

>   the lockfile from a subproject feels against the intent of GCV.

Reviewer: I think you need to clarify a bit more here. `GetVersionPlugin` is _part_ of GCV.

Author: nice have extended in 5b4be03ef8bd115122300ffef342c72597cb7b5e also noted that this approach doesnt work as the lock only contains external modules

- **Read the `gcvLocks` constraints instead of resolving.** The "without an exclusive

>   lock" error only applies to *resolution* — reading a configuration's declared

Reviewer: ```suggestion
- **Read the `gcvLocks` constraints instead of resolving.** The above error only applies to *resolution* — reading a configuration's declared
```

>   `DependencyConstraint`s is fine, so the no-config lookup could read the strict

Reviewer: can we drop the `no-config` bits throughout this doc pls

Author: 👍  fixed in 5b4be03ef8bd115122300ffef342c72597cb7b5e

  constraints off the root `gcvLocks` platform instead. This works for the common case
  but still reads the root project's model from a subproject, it needs special-casing for the
  `getVersion(project(...))` case.
- **Per-project view of the unified graph (chosen).** Each project resolves its *own*
  configuration instead of reaching into the root's `unifiedClasspath`. This is the
  isolated-projects-shaped route and is the only one that keeps all `getVersion`
  overloads working without special-casing.

## Decision

Each project resolves a per-project view of the unified dependency graph instead of
reaching into the root's `unifiedClasspath`.

> The trick is a **dependency-scope bucket** that both the resolvable (lock-computing) and

Reviewer: let's not introduce new nomenclature for things that have existing names. "bucket" in this case should be "configuration"

Reviewer: I'm also not a fan of "trick". This is how we're supposed to do stuff with configurations, we were just doing it wrong before/gradle moved on to a different system that we're adopting to now.

Author: removed the bucked reference 216df7d7ec08693b81ea430e1731c572c6ff2319

Author: fix the "trick" reference here 3d303e2bfdb90941e659a0255d67cb2d57546826

> consumable (per-project) views extend, so no configuration extends one of a different

Reviewer: this is not true? you can't extend configurations outside of your project.

Reviewer: https://docs.gradle.org/current/userguide/upgrading_version_8.html#extending_configurations_in_same_project

Author: whoops I meant root-project, fixed and made clearer (i hope) in b6be120b66d6ee7d7dbfdcdc2bd00fe0d5b8c0b9

role (Gradle 9 warns on a consumable extending a resolvable, and vice versa).

> - **`VersionsLockPlugin`** (root): collects every project's production + test deps into a

Reviewer: deps -> dependencies

Author: sorry I have such a habit of shortening thing! fixed in b2411a72d361f998cfe846e691bce3cf94b007fc

  new dependency-scope bucket, `unifiedClasspathDependencies`. `unifiedClasspath`
  (resolvable) now just `extendsFrom` the bucket and still computes the lock state —

>   behaviour is unchanged, the bucket is purely an intermediate.

Reviewer: The phrasing here still feels like we're describing a trick, when in actual fact it's what we're supposed to do.

> the bucket is purely an intermediate

This is essentially fluff, it's the entire purpose for dependency scope configurations, so either should be assumed or described as such.

Author: fixed in c9b9a1b5aabbe4bc08c1eacd13aba7fc1416243c

> - **`GetVersionPlugin`** (root): adds `gcvGetVersionElements`, a *consumable* view that

Reviewer: reading this more, I don't think it makes sense for `GetVersionPlugin` to add this? 

Typically, what I've seen is that you have:
* name of thing that was resolvable e.g. `runtimeClasspath`
* name of thing that was the dependency scope `implementation`
* the consumable version of the thing being suffixed with `Elements` e.g. `runtimeElements`

I think a similar approach makes sense here. i.e. you have:

* `unifiedClasspathDependencies` for the declarable configuartion
* `unifiedClasspath` for the resolvable - we can't change this for obvious reasons
* `unifiedClasspathElements` for the consumable.

The idea is that you want to have a consumable view of the `unifiedClasspath`, it's not restricted to just the `getVersion`s workflow, it just so happens to be the initial and only consumer (for now).

Author: good idea! then we wouldn't need two `GetVersionPlugin`s - fixed in a0fa53e2644c06ac261fb74e641b0511a75fba85

>   also extends the bucket, carrying capability `gcv:get-versions:0` and the `GCV_SOURCE`

Reviewer: don't think the `GCV_SOURCE` is worth mentioning

Author: :+1 fixed in c9b9a1b5aabbe4bc08c1eacd13aba7fc1416243c

  usage. Applies `GetVersionProjectPlugin` to every project.
- **`GetVersionProjectPlugin`** (new, per-project): registers a resolvable
  `gcvGetVersions` that depends on `project(':')` requesting that capability, and wires

>   the `getVersion(...)` extension to resolve that configuration.

Reviewer: ```suggestion
- **`GetVersionProjectPlugin`** (new, per-project): registers a resolvable
  `gcvGetVersions` configuration that depends on `project(':')` requesting the same capability, and wires
  the `getVersion(...)` ? to resolve that configuration.
```

`getVersion` is not an extension?

```mermaid
flowchart TD

>     subgraph root["root project"]

Reviewer: * > every project's prod + test deps
   
    This should be: 

     > dependencies from every project's locked configurations

* your arrows are the wrong way in a number of places:
  * `gcvGetVersionElements` and `unifiedClasspath` both `extendsFrom` `unifiedClasspathDependencies`, and yet, the arrow seems to suggest the other way.
  * I would reverse the arrow between `getVersion` and `gcvGetVersions`, and add a label along the lines of `resolves`
  * from `gcvGetVersions` -> `gcvGetVersionElements`, I would instead say
     > depends on with capability gcv:get-versions:0
* > :child (and every project)
  
  Just say:

  > every subproject (including root)

* drop the bucket nomenclature everywhere, these are configurations not buckets! 
* It would be good to colour code the boxes, i.e. what's dependency scope, what's resolvable, what's consumable, what's a just a normal function. See the official gradle [docs](https://docs.gradle.org/current/userguide/java_library_plugin.html#resolvable_configurations) for the java library plugin that describes this.

Author: between f3516b9d0b471ad4a8e0b35c0a5b567ac05a2e4e and 556d4e3e12ea42773f0ce74171dfa6637cd52590 should have it better

Reviewer: nice!

        bucket["<b>unifiedClasspathDependencies</b><br/><i>dependency-scope bucket</i><br/>(every project's prod + test deps)"]
        elements["<b>gcvGetVersionElements</b><br/><i>consumable</i> · cap gcv:get-versions:0"]
        unified["<b>unifiedClasspath</b><br/><i>resolvable</i> → writeLocks / verifyLocks"]
        bucket -->|extendsFrom| elements
        bucket -->|extendsFrom| unified
    end
    subgraph child[":child (and every project)"]
        getv["<b>gcvGetVersions</b><br/><i>resolvable</i>"]
        gv["getVersion('g:n')<br/><i>reads its own resolution result</i>"]
        getv --> gv
    end
    getv -->|"project(':') + requireCapabilities"| elements

    style root fill:none,stroke:#bbb,color:#111
    style child fill:none,stroke:#bbb,color:#111
```

## Consequences

Because `getVersion` resolves the unified graph in the *calling* project, the result
still carries both locked external versions and in-build project versions, and it never
resolves a cross-project configuration — so the exclusive-lock error is gone and the

> `getVersion('group:name', configuration)` / `getVersion(project(...))` cases keep working

Reviewer: > resolves the unified graph in the *calling* project

* what is the unified graph?
* if you're referring to the `unifiedClasspath`, it's a little weird to say it's resolved in the *calling* project, the whole premise is that you're not resolving that specific configuration in the calling project.

> `getVersion('group:name', configuration)` / `getVersion(project(...))` cases keep working

This is the first time you're mentioning these overloads, why do these matter and why is it a corollary to the work you've just done. Genuinely asking because I don't see how it follows

Author: I was referencing all the dependencies that are stored in the new dependency scoped configuration  that the consumable extends as didn't want to use `unifiedClasspath` when thats not what we are actually resolving - will make this more explicit

> first time you're mentioning these overloads

I wanted to make it clear that these had not changed (which isn't true anymore 9ed01041f682c8051b7889394a701d4972745143) will update to reflect how it actually is!

Author: hopefully I made this clearer here 9398402cfdfb646056878b788d0e482e64f21c72
